samples/MQTTSN/src/paho_iot_hub_telemetry_example_qos_0.c

```diff
@@ -0,0 +1,265 @@
+  }
+
+  return AZ_OK;
+}
+
+static az_result read_configuration_and_init_client()
+{
+  az_span device_id_span = AZ_SPAN_FROM_BUFFER(device_id);
+  AZ_RETURN_IF_FAILED(read_configuration_entry(
+      ENV_DEVICE_ID, ENV_DEVICE_ID, "", false, device_id_span, &device_id_span));
+
+  az_span iot_hub_hostname_span = AZ_SPAN_FROM_BUFFER(iot_hub_hostname);
+  AZ_RETURN_IF_FAILED(read_configuration_entry(
+      ENV_IOT_HUB_HOSTNAME,
+      ENV_IOT_HUB_HOSTNAME,
+      "",
+      false,
+      iot_hub_hostname_span,
+      &iot_hub_hostname_span));
+
+  // Initialize the hub client with the hub host endpoint and the default connection options
+  AZ_RETURN_IF_FAILED(az_iot_hub_client_init(
+      &client,
+      az_span_slice(iot_hub_hostname_span, 0, (int32_t)strlen(iot_hub_hostname)),
+      az_span_slice(device_id_span, 0, (int32_t)strlen(device_id)),
+      NULL));
+
+  return AZ_OK;
+}
+
+// This sample implements QoS 0
+int main(int argc, char** argv)
+{
+	int rc = 0;
+	int mysock;
+	unsigned char buf[500];
+	int buflen = sizeof(buf);
+	MQTTSN_topicid topic;
+	MQTTSNString topicstr;
+	int len = 0;
+	int retained = 0;
+	
+	// Default
+	char *host = "127.0.0.1";
+	int port = 10000; // use unicast port if sending a unicast packet
+
+	MQTTSNPacket_connectData options = MQTTSNPacket_connectData_initializer;
+	unsigned short topicid;	
+
+	// Create a unicast UDP socket
+	mysock = transport_open();
+	if(mysock < 0)
+		return mysock;
+
+	// Read for optional destination address and port
+	if (argc > 1)
+		host = argv[1];
+
+	if (argc > 2)
+		port = atoi(argv[2]);
+
+	printf("Connecting to hostname %s port %d\n", host, port);
+
+	// Read in the necessary environment variables and initialize the az_iot_hub_client
+	if (az_failed(rc = read_configuration_and_init_client()))
+	{
+		printf("Failed to read configuration from environment variables, return code %d\n", rc);
+		return rc;
+	}
+
+	options.clientID.cstring = device_id;
+
+	// Get topic name that the IoT Hub is subscribed to
+	if (az_failed(
+		rc = az_iot_hub_client_telemetry_get_publish_topic(
+			&client, NULL, topicname, sizeof(topicname), NULL)))
+	{
+		printf("Failed to get publish topic, return code %d\n", rc);
+		return rc;
+	}
+	
+	// CONNECT to MQTTSN Gateway
+	len = MQTTSNSerialize_connect(buf, buflen, &options);
+
+	if (az_failed(
+		rc = transport_sendPacketBuffer(host, port, buf, len)))
+	{
+		printf("Failed to send Connect packet to the Gateway, return code %d\n", rc);
+		return rc;
+	}
+	
+	// Wait for CONNACK from the MQTTSN Gateway
+	if (MQTTSNPacket_read(buf, buflen, transport_getdata) == MQTTSN_CONNACK)
+	{
+		int connack_rc = -1;
+
+		if (MQTTSNDeserialize_connack(&connack_rc, buf, buflen) != 1 || connack_rc != 0)
+		{
+			printf("Failed to receive Connect ACK packet, return code %d\nExiting...\n", connack_rc);
+			goto exit;
+		}
+		else 
+			printf("CONNACK rc %d\n", connack_rc);
+	}
+	else
+	{
+		printf("Failed to connect to the Gateway\nExiting...\n");
+		goto exit;
+	}

```

Comment:
Can we have small functions.
like 
connect
register
publish
subscribe
disconnect

This will help us to have better maintain this code

Reply:
and need retry logic (see comments above)

Reply:
do we have subscribe in our case? for now, I just have:

initialize
connect_device
register_topic
send_telemetry
disconnect_device

Reply:
for now leave the subscribe